### ISSUE TITLE
Inline jsdoc types do not correctly mark an argument as optional

### ISSUE THREAD

**Issue by jespertheend** [microsoft/TypeScript#51110]
# Bug Report

### 🔎 Search Terms

jsdoc inline optional 

### 🕗 Version & Regression Information

- This is the behavior in every version I tried, and I reviewed the FAQ for entries about `jsdoc`

### ⏯ Playground Link

[Playground link with relevant code][1]

### 💻 Code

```ts
let vaguelyOptional = (/** @type {string} */ x) => {};

/**
 * @param {string} x
 */
let definitelyRequired = (x) => {};

vaguelyOptional = definitelyRequired;
vaguelyOptional();
```

### 🙁 Actual behavior

No type error at `vaguelyOptional = definitelyRequired` and no type error at `vaguelyOptional()` either.

### 🙂 Expected behavior

I would expect either one of the two last lines to emit a type error. Because right now it is possible to invoke a function without providing an argument, even though that function depends on the argument.

In #47653 it was decided that using inline jsdoc is not meant to mark an argument as required. (To be honest I don't agree with this but that's a discussion for another time).
So it seems like in this case the expected behaviour would be to emit a type error at the `vaguelyOptional = definitelyRequired`.

More explanation and a better example can be found at the [Playground link][1]

[1]: https://www.typescriptlang.org/play?strictNullChecks=true&strictPropertyInitialization=false&strictBindCallApply=false&noImplicitThis=false&noImplicitReturns=false&noUncheckedIndexedAccess=true&alwaysStrict=false&noUnusedLocals=true&useUnknownInCatchVariables=false&exactOptionalPropertyTypes=false&noFallthroughCasesInSwitch=false&noImplicitOverride=false&noPropertyAccessFromIndexSignature=false&ts=4.8.4&filetype=js&noUnusedParameters=true#code/DYUwLgBAbghg5gVxMAngeQA5gJYHsB2MwEAvBABQD0AVNRAAJgoYgQDeAzmAE7b5wBfCNUoQAHgEpSAPnYCA3AChFNaouEMMMbjAC27Lr35Cx6kYtCQAJiABmfbGGQoASiACOCbNxBXSFSRk5JRVaM01tPQMePkEIAG0xAF0zSgtwCBt7fEdnTBwCIn9yQJJZNgVlSlEAFQALbA4IEF1HJph8Zu5uXG4AGggYJpAxFgBjJysAOgh61jAAd1wIDmw4QjAEHw4VUSsCAHJIOpgoebrWDj155hApxSyHJ1R8vEJiMkec59cPLx8rCFqrMGk19iAmvhcJAWo5Bp0QN1egMYJB7NwuBAwKCViAWk0xr0fBMVnwxiBdhADrBEHksG8iAdMrgIfgjriWljliczlSxEyhvCILh6YViFodLpwIj7jSkC9Re9-F9cqg3J5vL4gaIABK4BYgM79LE4hYwFAQFC4BAQVpwOqQbgITp8MDLRocJBNYCoxHC-ADABGIDGMAQHApwLhjQgGFwHFWgdAEFsvSpcrpBXeTLdEGDEFDwFAfgWjjq1rAlJgsciUqc3AgpexFfhFqYLC6PQxAC5FBmFVmiOQJNr0-B5ehFYz4X4Diqfur-r4mYHoXUOfpczzzpHRFcpaT1qitqwFhdOuWjbEBoaQJ1mwh7SbWD6wPXG9gi5TsT0FoMsbcnZpmed4QHwUC4AA1r4H7Yv+rQJrENaStK3C9vOziLpqVjDvIQA

**Comment by jespertheend** [comment]
For the record, in the playground link I have enabled `strictFunctionTypes`. Without this I wouldn't expect a type error to be emitted.

**Comment by jespertheend** [comment]
I've found a similar case, not sure if this is the same issue but I think so.
Either way, with this one `strictFunctionTypes` is not required to be set: [playground](https://www.typescriptlang.org/play?strictNullChecks=true&strictFunctionTypes=false&strictPropertyInitialization=false&strictBindCallApply=false&noImplicitThis=false&noImplicitReturns=false&noUncheckedIndexedAccess=true&alwaysStrict=false&noUnusedLocals=true&useUnknownInCatchVariables=false&exactOptionalPropertyTypes=false&noFallthroughCasesInSwitch=false&noImplicitOverride=false&noPropertyAccessFromIndexSignature=false&ts=4.8.4&filetype=js&noUnusedParameters=true#code/PQKhCgAIUgBAHAhgJ0QW0gbwM4BdkCWAdgOaQA+kRArmgEYCmyAvpAB5QjDgBm1RAY1wEA9kUh5CpAHK0AFGwCUWKJDVsAdLhEBlfMRJzFAbnDNwF8KAjQ4SVBhz7SrDtG4AbBrkjIGAR2oCPwATPSkyAF5IBWVIgD4sZlM-QOCGMOcoiSzZNFNgYEgAFQALAmxIJmQRZGwAGkgAd3KBUsgKyBIRERD6qyKRXFKmJoqGZommkWoPEK7vSERffmE0CeragC5wVKDQ8IMjU0srME47FHQsSQNWAG02AF1OT0WReGExRA9D0khorEAYlMMkBiVypUftgRFVkDU6ktKgw2PAGEIMjsPl8iD8-tlbjJaKZsaJcb8ssdTl4fAA3RAkagMDwATwA8p8yT8ATFrHBcCy0TcsqwuOw4iCwfTGcz2Zzvh4eYSSHlTNKmayOTiflTCpBpLDNshICM-I0YZMqkR5tR4M0CMMlisiGsNvDapAtnIgA)

**Comment by sandersn** [comment]
The rule is that *calls* in js don't enforce arity, which means that the last line doesn't error. The signatures both list their parameters as required, which means that second-to-last line doesn't error.
The difference between `@param` and inline per-parameter `@type` is not particularly consistent but I believe that direct annotation in the style of `vaguelyOptional` is rare, so it hasn't been worthwhile to add `@type` annotations to the rule.

I marked this bug Working As Intended because I believe it's better for the compiler to nudge people towards the usual `@param/@return` block comment. If you have a really solid use case for per-param `@type`, that could be enough to switch the bug to Help Wanted.

**Comment by jespertheend** [comment]
Tbh I can't remember my exact use case, but the reason why I was using inline param types was that I thought it was less verbose.

For example, deno/std [has a `spy` function](https://github.com/denoland/deno_std/blob/0ce558fec1a1beeda3ed7710c6507b9e991829c9/testing/mock.ts#L321-L325) that has generic parameters that takes the signature of the passed in function.
```js
const spyFn = spy((/** @type {number} */ x) => x);
```
In this case spyFn automatically gets arguments `[x?: number]` and return type `number`.

Doing it this way is a lot less verbose than
```js
/**
 * @param {number} x
 */
const fn = x => x;
const spyFn = spy(fn);
```

But now that I look at it, this doesn't seem that much more verbose so if it stays this way that's fine by me.

I think the biggest issue is that I ran into a runtime error after doing this, and it took me a couple of minutes to figure out why. I'm not sure if I will run into this again in the future, now that I know about this behaviour. So fixing this would mostly be beneficial for new users I think.
